### PR TITLE
[NTOS:EX] QSISystem*HandleInformation(): Fix number of handles

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

Do not increase number of handles, when locking a handle failed.
Plus related optimizations.

JIRA issue: [CORE-11978](https://jira.reactos.org/browse/CORE-11978)

## Proposed changes

- Set number of handles once only, with correct value.
- Set error status once only, when buffer is too small.

And while there:
- Require a full struct, on input.
- Return 0 as required size, if input buffer size is less than minimal.

### CODE REVIEW

File: ntoskrnl/ex/sysinfo.c [hunk 1]
```diff
@@ -1362,7 +1356,9 @@ QSI_DEF(SystemHandleInformation)
                     /* Check user's buffer size */
                     if (*ReqSize > Size)
                     {
+                        ExUnlockHandleTableEntry(HandleTable, HandleTableEntry);
                         Status = STATUS_INFO_LENGTH_MISMATCH;
+                        goto Abort;
```
Comment (HBelusca) [inline ntoskrnl/ex/sysinfo.c:1361]: Please write an apitest that shows whether this is the behaviour happening on Windows, namely, whether `HandleInformation->NumberOfHandles` will return the number of handles stored in the user buffer, or will return the number of handles in the system, regardless of the size of the user buffer.

**Review by HBelusca — DISMISSED** [review]

File: ntoskrnl/ex/sysinfo.c (hunk 1 above)
Comment (SergeGautherie) [reply ntoskrnl/ex/sysinfo.c:1361]: Adding that shortcut was wrong.
I'm working on extending ntdll:info winetest, separately.

File: ntoskrnl/ex/sysinfo.c [hunk 2]
```diff
@@ -1303,12 +1303,15 @@ QSI_DEF(SystemHandleInformation)
 
     DPRINT("NtQuerySystemInformation - SystemHandleInformation\n");
 
-    /* Set initial required buffer size */
-    *ReqSize = FIELD_OFFSET(SYSTEM_HANDLE_INFORMATION, Handles);
+    /* Set minimal input buffer size */
+    *ReqSize = sizeof(SYSTEM_HANDLE_INFORMATION);
```
Comment (HBelusca) [inline ntoskrnl/ex/sysinfo.c:1307]: What happens on Windows if you call `NtQuerySystemInformation` for that `SystemHandleInformation` class, with the `ReqSize` pointer parameter being NULL?

File: ntoskrnl/ex/sysinfo.c [hunk 3]
```diff
@@ -1350,21 +1353,14 @@ QSI_DEF(SystemHandleInformation)
             if ((HandleTableEntry->Object) &&
                 (HandleTableEntry->NextFreeTableEntry != -2))
             {
-                /* Increase of count of handles */
-                ++HandleInformation->NumberOfHandles;
```
Comment (HBelusca) [inline ntoskrnl/ex/sysinfo.c:1354]: Are you sure this gets incremented only if `ExpLockHandleTableEntry` succeeds? (you're now using the `Index` value to hold the number)

File: ntoskrnl/ex/sysinfo.c [hunk 4]
```diff
@@ -1391,12 +1387,13 @@ QSI_DEF(SystemHandleInformation)
...
                             HandleTableEntry->GrantedAccess;
-
-                        ++Index;
                     }
 
                     /* Unlock it */
                     ExUnlockHandleTableEntry(HandleTable, HandleTableEntry);
+
+                    /* Increase count of handles, always */
+                    ++Index;
```
Comment (HBelusca) [inline ntoskrnl/ex/sysinfo.c:1396]: `Index` is exclusively used for the index into the `Handles` array of the user-provided `HandleInformation` buffer. Please keep the original intent of the code and separate that from the counting of handles (the `NumberOfHandles` variable) that is also done separately, whatever the size of the user-provided information buffer.

**Comment by HBelusca** [comment]
As part of the tests, please verify that the sysinternals Handle utility: https://docs.microsoft.com/en-us/sysinternals/downloads/handle , as well as the `openfiles` Windows utility, still works on ReactOS.

File: ntoskrnl/ex/sysinfo.c [hunk 5]
```diff
@@ -1303,12 +1303,15 @@ QSI_DEF(SystemHandleInformation)
...
 
-    /* Set initial required buffer size */
-    *ReqSize = FIELD_OFFSET(SYSTEM_HANDLE_INFORMATION, Handles);
+    /* Set minimal input buffer size */
+    *ReqSize = sizeof(SYSTEM_HANDLE_INFORMATION);
 
     /* Check user's buffer size */
     if (Size < *ReqSize)
     {
+#if (NTDDI_VERSION < NTDDI_VISTA)
```
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply ntoskrnl/ex/sysinfo.c:1312]: Yes.

File: ntoskrnl/ex/sysinfo.c (hunk 2 above)
Comment (SergeGautherie) [reply ntoskrnl/ex/sysinfo.c:1307]: Nothing special: that parameter is optional. (See #3910.)

File: ntoskrnl/ex/sysinfo.c (hunk 3 above)
Comment (SergeGautherie) [reply ntoskrnl/ex/sysinfo.c:1354]: See Jira issue:
I am not 100% sure, as I do not know how to test this very case.
Nonetheless, it feels more self-consistent, and I assume current winetest was not reported as (intermittently) failing.

File: ntoskrnl/ex/sysinfo.c (hunk 4 above)
Comment (SergeGautherie) [reply ntoskrnl/ex/sysinfo.c:1396]: I will revert that, though the 2 variables feel redundant after the counting change.

**Review by SergeGautherie — COMMENTED** [review]
> sysinternals Handle utility [...] `openfiles` Windows utility

I can try: what (steps) should I check more precisely?

File: ntoskrnl/ex/sysinfo.c (hunk 2 above)
Comment (HBelusca) [reply ntoskrnl/ex/sysinfo.c:1307]: Precisely, and this is why here, if the user sets `ReqSize` to be NULL, now here and in the other places below (and in the original code as well) there would be a NULL pointer dereference.

File: ntoskrnl/ex/sysinfo.c (hunk 3 above)
Comment (GeoB99) [reply ntoskrnl/ex/sysinfo.c:1354]: When it comes to kernel development, you're either sure of yourself of what you're doing or you aren't. Assumptions are to be avoided.

If you're unsure then it's at best not worth touching the parts of the kernel you're clearly not fond with.

**Review by GeoB99 — CHANGES_REQUESTED** [review]
When it comes to kernel development, you're either sure of yourself of what you're doing or you aren't. Assumptions are to be avoided.

If you're unsure then it's at best not worth touching the parts of the kernel you're clearly not fond with.

**Pull request closed without merging**